plugins/bkpr/descriptions.c

```diff
@@ -0,0 +1,323 @@
+{
+	struct utxo_desc *ud;
+
+	ud = utxo_desc_htable_get(descriptions->by_utxo, outpoint);
+	if (ud) {
+		tal_free(ud->desc);
+		ud->desc = tal_strdup(ud, desc);
+	} else {
+		ud = tal(descriptions->by_utxo, struct utxo_desc);
+		ud->outp = *outpoint;
+		ud->desc = tal_strdup(ud, desc);
+		utxo_desc_htable_add(descriptions->by_utxo, ud);
+	}
+	return ud;
+}
+
+void add_utxo_description(struct command *cmd,
+			  struct bkpr *bkpr,
+			  const struct bitcoin_outpoint *outpoint,
+			  const char *desc TAKES)
+{
+	struct utxo_desc *ud;
+
+	ud = new_utxo_description(bkpr->descriptions, outpoint, desc);
+	utxo_desc_datastore_update(cmd, ud);
+}
+
+static void memleak_scan_utxo_desc_htable(struct htable *memtable,
+					  struct utxo_desc_htable *ht)
+{
+	memleak_scan_htable(memtable, &ht->raw);
+}
+
+static void memleak_scan_payment_hash_desc_htable(struct htable *memtable,
+						  struct payment_hash_desc_htable *ht)
+{
+	memleak_scan_htable(memtable, &ht->raw);
+}
+
+/* To avoid JSON encoding/decoding, we use hex encoding directly */
+static char *json_cstr_from_hex(const tal_t *ctx, const char *buffer, const jsmntok_t *tok)
+{
+	char *result;
+	size_t hexlen, rawlen;
+	hexlen = tok->end - tok->start;
+	rawlen = hex_data_size(hexlen);
+
+	result = tal_arr(ctx, char, rawlen + 1);
+	if (!hex_decode(buffer + tok->start, hexlen, result, rawlen))
+		return tal_free(result);
+
+	result[rawlen] = '\0';
+	return result;
+}
+
+struct descriptions *init_descriptions(const tal_t *ctx,
+				       struct command *init_cmd)
+{
+	struct descriptions *descriptions = tal(ctx, struct descriptions);
+	struct json_out *params;
+	const jsmntok_t *result, *datastore, *t;
+	size_t i;
+	const char *buf;
+
+	descriptions->by_utxo = tal(descriptions, struct utxo_desc_htable);
+	utxo_desc_htable_init(descriptions->by_utxo);
+	memleak_add_helper(descriptions->by_utxo, memleak_scan_utxo_desc_htable);
+
+	descriptions->by_payment_hash = tal(descriptions, struct payment_hash_desc_htable);
+	payment_hash_desc_htable_init(descriptions->by_payment_hash);
+	memleak_add_helper(descriptions->by_payment_hash, memleak_scan_payment_hash_desc_htable);
+
+	/* Load everything under "bookkeeper/description/utxo" */
+	params = json_out_new(tmpctx);
+	json_out_start(params, NULL, '{');
+	json_out_start(params, "key", '[');
+	json_out_addstr(params, NULL, "bookkeeper");
+	json_out_addstr(params, NULL, "description");
+	json_out_addstr(params, NULL, "utxo");
+	json_out_end(params, ']');
+	json_out_end(params, '}');
+
+	result = jsonrpc_request_sync(tmpctx, init_cmd,
+				      "listdatastore",
+				      params, &buf);
+
+	datastore = json_get_member(buf, result, "datastore");
+	json_for_each_arr(i, t, datastore) {
+		const jsmntok_t *keytok, *hextok;
+		const char *desc_str;
+		struct bitcoin_outpoint outp;
+
+		keytok = json_get_member(buf, t, "key");
+		if (keytok->size != 4)
+			continue;
+
+		/* ["bookkeeper", "description", "utxo", <key>] */
+		hextok = json_get_member(buf, t, "hex");
```

Comment:
where is the `hex` keyword in the datastore struct that's returned? does every entry return a `key` and a `hex`?

Reply:
Yes, we will always have it on the final key (there won't be any on directories).